67 feature implement pure pursuit controller

schwalga opened the pull request:
# Description

This PR implements the pure pursuit controller. The first tests suggest, that it is working as intended, however due to problems with the GPS signal the car is not able to hold course at the moment.

While i tried fixing the problem at first under this issue, after a few hours of trial and error I realise, that this warrants [a seperate issue](#131). The documentation added in this PR gives a first overview over the problem it will be amended with the final filter in #131. For now I have left the simple filter added to `DummyTrajectorySub.py` so that the controller can be tested by a reviewer. 

Fixes #67
## Time invested

- 16h @schwalga (Most of the work has been done in 66-test due to a misunderstanding (see cc59a2e)) 
- 7h @Julian-Graf 

## Type of change

- New feature (non-breaking change which adds functionality)

## Does this PR introduce a breaking change?

- No

## Most important changes

The most important file is `pure_pursuit_controller.py`

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works (might be obsolete with CI later on)
- [x] New and existing unit tests pass locally with my changes (might be obsolete with CI later on)

JoKircher reviewed (changes requested):
Nice new functionality for the action portion of the agent.

Unfortunately, due to an issue with the docker container. I couldn't test the controller so far. There are still quite some todos so the PR is blocked for now.

If you have any questions please let me know!

One last note:

The localization you implemented will probably need at different corners of the project, so a refactor to a more centralized spot might be a good idea in the future.

Julian-Graf commented on code/acting/src/acting/pure_pursuit_controller.py:
```diff
@@ -0,0 +1,261 @@
+            f"/carla/{self.role_name}/pure_pursuit_steer",
+            qos_profile=1)
+
+        self.status_pub = self.new_publisher(  # todo: delete (this is only
+            # necessary if vehicle controller isn't running)
+            Bool, f"/carla/{self.role_name}/status",
+            qos_profile=QoSProfile(
+                depth=1,
+                durability=DurabilityPolicy.TRANSIENT_LOCAL)
+        )
```
This publisher has to be deleted before merging

schwalga replied:
Fixed in 426f8db

Julian-Graf commented on code/acting/src/acting/pure_pursuit_controller.py:
```diff
@@ -0,0 +1,261 @@
+        :return:
+        """
+        self.loginfo('PurePursuitController node running')
+        self.status_pub.publish(True)  # todo: delete this is only
+        # necessary if vehicle controller isn't running
```
This has to be deleted before merging

schwalga replied:
Fixed in 426f8db

Julian-Graf commented on code/acting/src/acting/pure_pursuit_controller.py:
```diff
@@ -0,0 +1,261 @@
+        self.__velocity = data.speed
+
+    def __calculate_steer(self) -> float:
+        l_vehicle = 2  # dist between front and rear wheels todo: measure
+        k = 10  # todo: tune
```
It makes sense to have the tuning as a separate issue

schwalga commented on code/acting/src/acting/DummyTrajectorySub.py:
```diff
@@ -106,12 +107,17 @@ def output_average_gps_2_xyz(self, data: NavSatFix):
-            x1 = self.pos_average[0] / 5
-            y1 = self.pos_average[1] / 5
-            z1 = self.pos_average[2] / 5
+        w: float = 0.75  # weight of the new position compared to the old
+        avg: int = 10
```
Mostly trial-and-error. `gps_signal__example.md` contains some of these results.

schwalga commented on code/acting/src/acting/pure_pursuit_controller.py:
```diff
@@ -0,0 +1,261 @@
+
+        def loop(timer_event=None):
+            """
+            Main loop of the acting node
```
Stanley and pure pursuit are both running at the same time, the vehicle controller subscribes to both controllers and uses the one it deems more suitable at the moment.
The documentation for the loop is obviously incorrect .

schwalga replied:
Fixed in 426f8db

schwalga commented on code/acting/src/acting/DummyTrajectorySub.py:
```diff
@@ -94,6 +94,7 @@ def output_average_gps_2_xyz(self, data: NavSatFix):
         The current position is then updated and published
         in the PoseStamped format.
         :param data: message according to NavSatFix definition
+        :param average: number of points to be averaged
```
This parameter is a relic from previous iterations, however this entire functionality will be moved/redone in #131.

schwalga replied:
Fixed in 426f8db

schwalga commented on code/acting/src/acting/helper_functions.py:
```diff
@@ -0,0 +1,94 @@
+from scipy.spatial.transform import Rotation
+
+
+# todo: docs
```
Fixed in 426f8db

schwalga commented on code/acting/src/acting/helper_functions.py:
```diff
@@ -0,0 +1,94 @@
+
+
+# todo: docs
+def vectors2angle(x1: float, y1: float, x2: float, y2: float) -> float:
```
Fixed in 426f8db

schwalga commented on code/acting/src/acting/helper_functions.py:
```diff
@@ -0,0 +1,94 @@
+    return math.degrees(theta)
+
+
+def quaternion2heading(x: float, y: float, z: float, w: float) -> float:
```
Fixed in 426f8db

schwalga commented on code/acting/src/acting/helper_functions.py:
```diff
@@ -0,0 +1,94 @@
+        angle (float): The angle to normalize
+    Returns:
+        float: Angle in radian in [-pi, pi]
+    """
```
Fixed in 426f8db

schwalga commented on code/acting/src/acting/helper_functions.py:
```diff
@@ -0,0 +1,94 @@
+        pose (Pose): The current pose of the ego vehicle
+    Returns:
+        float: The normalized yaw
+    """
```
Fixed in 426f8db

schwalga commented on code/acting/src/acting/helper_functions.py:
```diff
@@ -0,0 +1,94 @@
+        idx (int): The pose index
+    Returns:
+        float: [description]
+    """
```
Fixed in 426f8db

schwalga commented on doc/04_examples/gps_example/gps_signal_example.md:
```diff
@@ -0,0 +1,59 @@
+07.01.2023
+
+---
+[TOC]
```
Fixed in 426f8db

schwalga commented on doc/04_examples/gps_example/gps_signal_example.md:
```diff
@@ -0,0 +1,59 @@
+### Intuitive filter
+
+The first way combines the two most intuitive methods of smoothing the input.
+At first, multiple ($m$) inputs are summed up and divided to take an average.
```
I can't find the reason why this is not working...

schwalga replied:
Fixed in 58f7565

schwalga commented on doc/04_examples/gps_example/gps_signal_example.md:
```diff
@@ -0,0 +1,59 @@
+The number of separate inputs taken into account for the average and the weight of old input can be seen as tweak-able
+parameters.
+The following graphs were taken while the car was stationary, the time on the bottom is therefore irrelevant.
+Shown is the position translated to a local coordinate system, the transformation will be discussed in //TODO add ref.
```
Fixed in 426f8db

schwalga commented on code/acting/src/acting/pure_pursuit_controller.py:
```diff
@@ -0,0 +1,261 @@
+
+
+# todo: docs
+# todo: test
```
Fixed in 426f8db

schwalga commented on code/acting/src/acting/pure_pursuit_controller.py:
```diff
@@ -0,0 +1,261 @@
+            self.__heading = 0
+            return
+        cur_x = self.__position[0] - self.__last_pos[0]
+        cur_y = self.__position[1] - self.__last_pos[1]
```
Fixed in 426f8db

schwalga commented on code/acting/src/acting/pure_pursuit_controller.py:
```diff
@@ -0,0 +1,261 @@
+        target_v_x = target_wp.pose.position.x - self.__last_pos[0]
+        target_v_y = target_wp.pose.position.y - self.__last_pos[1]
+        cur_v_x = self.__position[0] - self.__last_pos[0]
+        cur_v_y = self.__position[1] - self.__last_pos[1]
```
Fixed in 426f8db

schwalga commented on doc/04_examples/gps_example/gps_signal_example.md:
```diff
@@ -0,0 +1,59 @@
+of the output signal.
+
+![GPS signal (m=1, w=0,5)](../../00_assets/filter_img/avg_10_w_1_000.png)
+Using a large number of data points ($m = 10$) also improves the magnitude of the noise.
```
Fixed in 58f7565

schwalga commented:
I'm not sure how changing the md document managed to break the build?
Apart from that, I fixed all the issues mentioned in the review (ca. $1 \frac{1}{2}$ h by @schwalga)

JoKircher commented:
For some reason, the leaderboard evaluation failed. Maybe just rerunning the tests should suffice otherwise @nylser might be of help.

nylser commented:
> For some reason, the leaderboard evaluation failed. Maybe just rerunning the tests should suffice otherwise @nylser might be of help.
Looks like this might be an issue because I didn't get to switching it to the right machine yet? VK_DEVICE_LOST is the error code, which might hint to the GPU already being used differently

JoKircher reviewed (commented):
Good job, all of the suggested changes have been incorporated. However, for now the ego vehicle drives to the right side of the road and hit a parking clock. I guess that is related to the GPS problem?

schwalga commented:
As suggested I have set the GPS noise to 0 for the time being. 
I found/fixed another issue, where points in the trajectory would be targeted even though they had already been reached. 
At the moment the car is capable of following the trajectory defined in `DummyTrajectoryPub.py`. 
Since some more problems have come to light, I think this PR is not ready to be merged.

schwalga commented:
fixed a few errors, tested a more parameters and added a PID controller for the steering angle.
I think this PR should be merged this sprint, so that others can test their features.

~ 6h by @schwalga

schwalga commented:
During testing with a more complex trajectory, a few errors appeared, which have now been fixed.
To test, simply enter b5 install and then b5 run, and the car will take the first turn :).
I would maybe clean up the code in a few places, but in my eyes the PR is ready for merging. 

I also added a custom message that useful  for #151  

~ 4h by @schwalga

JoKircher commented on code/acting/src/acting/pure_pursuit_controller.py:
```diff
@@ -0,0 +1,305 @@
+        self.debug_publisher.publish(debug_msg)
+        # <-
+
+        self.pure_pursuit_steer_target_pub.publish(target_wp.pose)
```
Is there a reason you use the `pose` but not the `poseStamped`?

schwalga replied:
I did not see the need for the added header. If this turns out to be a problem, it would be a quick fix later on.

JoKircher replied:
Alright, just have been curious

JoKircher reviewed (changes requested):
Nice PR, the pure pursuit controller seems to work nicely. One last question before I approve the PR. Is is intended to cut the corner at the left turn and drive in the middle of the two lanes afterwards?

Review time: 30 min

JoKircher commented:
Also could you please resolve the merge conflicts?

Julian-Graf commented on code/acting/src/acting/coordinate_transformation.py:
```diff
@@ -13,18 +13,18 @@
-    TOWN10 = 0, 0, 0
-    TOWN11 = 0, 0, 0
-    TOWN12 = 35.25000, -101.87500, 331.00000
+    TOWN01 = 0, 0, 0  # lat = 1.7e-08, lon = -5.6e-08, alt = 0.02
```
Where do these values come from? They are all pretty much 0

schwalga replied:
I tried to spawn at (0, 0, 0) on every map, and looked a the GPS sensor.
The only map where lat/lon was not equal to zero is town 12.

JoKircher reviewed (approved):
Looks good. Happily approve! Still some conflicts though
